### PR TITLE
feat: allow having controllers with a singleton lifetime and permit usage without scoped containers in koa state

### PR DESCRIPTION
**Pull request by dlgoodchild**

### CODE REVIEW

File: src/controller.ts [hunk 1]
```diff
@@ -19,24 +19,30 @@ export type ConstructorOrControllerBuilder =
   | (new (...args: Array<any>) => any)
   | IAwilixControllerBuilder
 
+export interface InstanceOptions {
+  singleton?: boolean
```
Comment (jeffijoe) [inline src/controller.ts:23]: Let's document these

File: src/invokers.ts [hunk 2]
```diff
@@ -89,7 +94,19 @@ export function makeResolverInvoker<T>(resolver: Resolver<T>) {
...
+        throw new Error('Awilix container not found on Koa state object. Please ensure you use either scopePerRequest or attachContainer')
+      }
+
+      if (singleton) {
+        if (!resolved) {
+          resolved = container.build(resolver)
+        }
+      }
+      else {
+        resolved = container.build(resolver)
```
Comment (jeffijoe) [inline src/invokers.ts:109]: Minor thing, but when using scoped, the last instantiated controller will be assigned to `resolved` which prevents it from being garbage collected until the next request comes in. Might not be a big deal, but could be solved with an intermediary variable.

File: src/scope-per-request.ts [hunk 3]
```diff
@@ -16,3 +16,22 @@ export function scopePerRequest(container: AwilixContainer) {
...
+ * Koa middleware factory that will simply attach the container
+ * to the context (ctx) state, with no additional scoping.
+ *
+ * You should only use one of either scopePerRequest or attachContainer.
+ *
+ * @param  {AwilixContainer} container
+ * @return {Function}
+ */
+export function attachContainer(container: AwilixContainer) {
+  return function scopePerRequestMiddleware(
```
Comment (jeffijoe) [inline src/scope-per-request.ts]:
```suggestion
  return function attachContainerMiddleware(
```

**Review by jeffijoe — CHANGES_REQUESTED** [review]
Minor things, looks good though!

**Comment by dlgoodchild** [comment]
Changes applied as suggested.

**Review by jeffijoe — APPROVED** [review]

**Comment by jeffijoe** [comment]
Thanks @dlgoodchild , nice work! 🙏

**Comment by dlgoodchild** [comment]
Thanks... happy to contribute!

**Pull request merged**